library/std/src/f16/div_euclid.rs

```diff
@@ -0,0 +1,111 @@
+#[cfg(test)]
+mod tests;
+
+/// Software implementation of `f16::div_euclid`.
+#[allow(dead_code)] // f16::div_euclid is cfg(not(test))
+pub(crate) fn div_euclid(a: f16, b: f16) -> f16 {
+    if let Some((a_neg, a_exp, a_m)) = normal_form(a)
+        && let Some((b_neg, b_exp, b_m)) = normal_form(b)
+    {
+        let exp = a_exp - b_exp;
+        match (a_neg, b_neg) {
+            (false, false) => div_floor(exp, a_m, b_m),
+            (true, false) => -div_ceil(exp, a_m, b_m),
+            (false, true) => -div_floor(exp, a_m, b_m),
+            (true, true) => div_ceil(exp, a_m, b_m),
+        }
+    } else {
+        // `a` or `b` are +-0.0 or infinity or NaN.
+        // `a / b` is also +-0.0 or infinity or NaN.
+        // There is no need to round to an integer.
+        a / b
+    }
+}
+
+/// Returns `floor((a << exp) / b)`.
+///
+/// Requires `2^10 <= a, b < 2^11`.
+fn div_floor(exp: i16, a: u16, b: u16) -> f16 {
+    if exp < 0 {
+        0.0
+    } else if exp <= 5 {
+        // aa < (2^11 << 5) = 2^16
+        let aa = a << exp;
+        // q < 2^16 / 2^10 = 2^6
+        let q = (aa / b) as u8;
+        // We have to use `as` because `From<u8> for f16` is not yet implemented.
+        q as f16
+    } else if exp <= 16 {
+        // aa < (2^11 << 16) = 2^27
+        let aa = u32::from(a) << exp;
+        let bb = u32::from(b);
+        // q < 2^27 / 2^10 = 2^17
+        let q = aa / bb;
+        q as f16
+    } else {
+        // exp >= 17
+        // result >= (2^10 << 17) / 2^11 = 2^16
+        // Exponent 16 is too large to represent in f16.
+        f16::INFINITY
+    }
+}
+
+/// Returns `ceil((a << exp) / b)`.
+///
+/// Requires `2^10 <= a, b < 2^11`.
+fn div_ceil(exp: i16, a: u16, b: u16) -> f16 {
+    if exp < 0 {
+        1.0
+    } else if exp <= 5 {
+        // aa < (2^11 << 5) = 2^16
+        let aa = a << exp;
+        // q < 2^16 / 2^10 + 1 = 2^6 + 1
+        let q = ((aa - 1) / b) as u8 + 1;
+        // We have to use `as` because `From<u8> for f16` is not yet implemented.
+        q as f16
+    } else if exp <= 16 {
+        // aa <= ((2^11 - 1) << 16) = 2^27 - 2^16
+        let aa = u32::from(a) << exp;
+        let bb = u32::from(b);
+        // q <= (2^27 - 2^16) / 2^10 + 1 = 2^17 - 2^6 + 1
+        let q = ((aa - 1) / bb) as u16 + 1;
```

Comment:
This panics with "attempt to add with overflow" given either of:

```rust
    div_ceil(16, 1248, 1248);
    div_euclid(-624.0, 0.009521484375);
```

Reply:
Yes that was a bug in f16::div_ceil, fixed.